crates/kumod/src/dmarc.rs

```diff
@@ -0,0 +1,106 @@
+use std::collections::BTreeMap;
+use std::net::SocketAddr;
+use std::str::FromStr;
+
+use config::{get_or_create_sub_module, serialize_options};
+use kumo_dmarc::{CheckHostParams, DmarcResult};
+use mailparsing::AuthenticationResult;
+use message::Message;
+use mlua::{Lua, LuaSerdeExt, UserDataRef};
+use serde::Serialize;
+
+use crate::smtp_server::ConnectionMetaData;
+
+#[derive(Debug, Serialize)]
+struct CheckHostOutput {
+    disposition: DmarcResult,
+    result: AuthenticationResult,
+}
+
+pub fn register<'lua>(lua: &'lua Lua) -> anyhow::Result<()> {
+    let dmarc_mod = get_or_create_sub_module(lua, "dmarc")?;
+
+    dmarc_mod.set(
+        "verify",
+        lua.create_async_function(
+            |lua,
+             (msg, dkim_result, meta): (
+                UserDataRef<Message>,
+                UserDataRef<Vec<AuthenticationResult>>,
+                UserDataRef<ConnectionMetaData>,
+            )| async move {
+                let addr = meta
+                    .get_meta("received_from")
+                    .and_then(|v| SocketAddr::from_str(v.as_str()?).ok())
+                    .expect("`received_from` is always set, and always to a value representing a `SocketAddr`");
+
+                let resolver = dns_resolver::get_resolver();
+
+                // MAIL FROM
+                let msg_sender = msg.sender();
+
+                let mail_from_domain = msg_sender.ok().map(|x| x.to_string());
```

Comment:
Probably want just the domain here?

```suggestion
                let mail_from_domain = msg_sender.ok().map(|x| x.domain().to_string());
```